module/os/linux/zfs/vdev_disk.c

```diff
@@ -834,38 +823,67 @@ vbio_completion(struct bio *bio)
  * NOTE: if you change this function, change the copy in
  * tests/zfs-tests/tests/functional/vdev_disk/page_alignment.c, and add test
  * data there to validate the change you're making.
- *
  */
 typedef struct {
-	uint_t  bmask;
-	uint_t  npages;
-	uint_t  end;
-} vdev_disk_check_pages_t;
+	size_t	blocksize;
+	int	seen_first;
+	int	seen_last;
+} vdev_disk_check_alignment_t;
 
 static int
-vdev_disk_check_pages_cb(struct page *page, size_t off, size_t len, void *priv)
+vdev_disk_check_alignment_cb(struct page *page, size_t off, size_t len,
+    void *priv)
 {
 	(void) page;
-	vdev_disk_check_pages_t *s = priv;
+	vdev_disk_check_alignment_t *s = priv;
+
+	/*
+	 * The cardinal rule: a single on-disk block must never cross an
+	 * physical (order-0) page boundary, as the kernel expects to be able
+	 * to split at both LBS and page boundaries.
+	 *
+	 * This implies various alignment rules for the blocks in this
+	 * (possibly compound) page, which we can check for.
+	 */
 
 	/*
-	 * If we didn't finish on a block size boundary last time, then there
-	 * would be a gap if we tried to use this ABD as-is, so abort.
+	 * If the previous page did not end on a page boundary, then we
+	 * can't proceed without creating a hole.
 	 */
-	if (s->end != 0)
+	if (s->seen_last)
 		return (1);
 
 	/*
-	 * Note if we're taking less than a full block, so we can check it
-	 * above on the next call.
+	 * TODO: Ideally we would require that len is a multiple of LBS.
+	 *       To do that we need to not allow setting ashift= smaller
+	 *       than the LBS. Often, it will end up failing this check
```

Comment:
ashift smaller than LBS should never happen, since attempt of ZFS write that would either fail or cause read-modify-write, if something below is expected to handle it, but we don't want it for several reasons.  If there is a way for ZFS to misread LUKS LBS, then it much be fixed, and user should not be allowed to set ashift smaller than LBS.  That is how it is on FreeBSD and how I believe it must be done.

Reply:
I observed that creating a 4K dm-crypt map, then doing `zpool create -o ashift=9 tank /dev/mapper/mycrypt` spewed zio errors. Looking at the `vdev_t` for the device, it had `vdev_logical_shift=12`, `vdev_physical_shift=12`, but `vdev_ashift=9`. As far as I can tell, `vdev_open` is responsible for setting these based on what it gets from the vdev driver. So, ZFS is not "misreading LUKS LBS", but something weird is happening further upstack, nothing to do with this.

If that was fixed, then I would do what I originally had in place of this TODO comment: making sure the data length within this page was an even multiple of the LBS.

Reply:
When the user provides an ashift it looks like we're only enforcing that it be `>= ASHIFT_MIN` and `<= ASHIFT_MAX`.  We should additionally check that it's `>= vdev_logical_shift`  and fail the `zpool create` if it's not.  This should do the trick, but I haven't had a chance to test it.

```diff
diff --git a/module/zfs/vdev.c b/module/zfs/vdev.c
index bcab46c63b..60a2e3e409 100644
--- a/module/zfs/vdev.c
+++ b/module/zfs/vdev.c
@@ -2222,6 +2222,7 @@ vdev_open(vdev_t *vd)
                        vd->vdev_attaching = B_FALSE;
                }
                if (vd->vdev_ashift != 0 && (vd->vdev_ashift < ASHIFT_MIN ||
+                   vd->vdev_ashift < vd->vdev_logical_ashift ||
                    vd->vdev_ashift > ASHIFT_MAX)) {
                        vdev_set_state(vd, B_TRUE, VDEV_STATE_CANT_OPEN,
                            VDEV_AUX_BAD_ASHIFT);
```

Reply:
I wonder if this is going to break for anyone? At least, I know that in the block layer block sizes are all converted to counts of 512b sectors, and it's up to drivers to enforce. I haven't checked specifically but I wouldn't be at all surprised to find that a great many drivers or drives don't actually care or know how to handle it.

I still think we should do it, but it feels like the kind of thing that needs a bunch of thinking first?

Reply:
I am really surprised that it does not work as I expected.  I don't think ashift property set on a pool should result in any error, rather it should be ignored at the time of vdev creation for vdevs where it makes no sense.  I'll take a look on it tomorrow, since today I am too tired and jet-lagged.  If though pool was created somehow created with ashift smaller than LBS (likely somebody messed up with VM configuration or some very weird disk driver) -- that must be fatal at the time of pool import.

Reply:
I think this should fix the issue: https://github.com/openzfs/zfs/pull/16690 .

Reply:
Nice. Meanwhile, I've removed the TODO comment and reintroduced the multiple-of-LBS check. See below.